src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -1240,5 +1240,154 @@ private static bool HasNamespaceName(NamespaceSymbol @namespace, string namespac
+            {
+                if (member.Kind != SymbolKind.Method)
+                {
+                    continue;
+                }
+                var method = (MethodSymbol)member;
+                if ((method.DeclaredAccessibility == Accessibility.Public) &&
+                    method.IsStatic &&
+                    (method.ParameterCount == 0) &&
+                    !method.IsGenericMethod)
+                {
+                    var returnType = method.ReturnType as NamedTypeSymbol;
+                    if ((object)returnType == null)
+                    {
+                        break;
+                    }
+                    createBuilderMethod = method;
+                    return returnType;
+                }
+            }
+            createBuilderMethod = null;
+            return null;
+        }
+
+        /// <summary>
+        /// Replace Task-like types with Task types.
+        /// </summary>
+        internal static TypeSymbol NormalizeTaskTypes(this TypeSymbol type, CSharpCompilation compilation)
+        {
+            NormalizeTaskTypesCore(compilation, ref type);
+            return type;
+        }
+
+        /// <summary>
+        /// Replace Task-like types with Task types. Returns true if there were changes.
+        /// </summary>
+        private static bool NormalizeTaskTypesCore(CSharpCompilation compilation, ref TypeSymbol type)
+        {
+            switch (type.Kind)
+            {
+                case SymbolKind.NamedType:
+                    var namedType = (NamedTypeSymbol)type;
+                    var changed = NormalizeTaskTypesCore(compilation, ref namedType);
+                    type = namedType;
+                    return changed;
+                case SymbolKind.ArrayType:
+                    // PROTOTYPE(tasklike): Use VisitType or similar to cover all cases.
+                    break;
+            }
+            return false;
+        }
+
+        private static bool NormalizeTaskTypesCore(CSharpCompilation compilation, ref NamedTypeSymbol type)
+        {
+            var constructedFrom = type.ConstructedFrom;
+            if ((object)constructedFrom == type)
+            {
+                MethodSymbol createBuilderMethod;
+                var builderType = type.GetAsyncMethodBuilderType(out createBuilderMethod);
+                if ((object)builderType == null)
+                {
+                    return false;
+                }
+                Debug.Assert(type.Arity < 2);
+                type = compilation.GetWellKnownType(
+                    type.Arity == 1 ?
+                    WellKnownType.System_Threading_Tasks_Task_T :
+                    WellKnownType.System_Threading_Tasks_Task);
+                return true;
+            }
+            else
+            {
+                var typeArgs = type.TypeArguments;
+                var builder = ArrayBuilder<TypeSymbol>.GetInstance();
+                bool hasChanged = false;
```

Comment:
I wonder if it'd perform better to run through just doing a check, and if it returns false then skip the builder.

Only allocate the builder in the (rare) case where it is tasklike. This will be slower in the non-Task tasklike case, but involve less memory allocation in the much more common non-tasklike case.